### PR TITLE
Fix filter typing and add expression type.

### PR DESCRIPTION
**Pull request by cns-solutions-admin**
See #1380.
Add typing for expressions and fix filter expressions.
improve filter conversion by adding typing and helping typescript to guess types.
add tests for filters mentioned in issue #1380.

no backports.

### CODE REVIEW

**Comment by HarelM** [comment]
Thanks for taking the time to do this!!
Any chance to add more tests? Seems like there are more code changes than tests changes, which I find odd.
Also, I'm not sure I understand the legacy part, can you elaborate on that?

File: build/generate-style-spec.ts [hunk 1]
```diff
@@ -128,36 +128,152 @@ export type ResolvedImageSpecification = string;
...
+    
+export type ExpressionSpecification = 
+    // types
+    | ['array', unknown | ExpressionSpecification] // array
+    | ['array', ExpressionInputType | ExpressionSpecification, unknown | ExpressionSpecification] // array
+    | ['array', ExpressionInputType | ExpressionSpecification, number | ExpressionSpecification, unknown | ExpressionSpecification] // array
+    | ['boolean', ...(unknown | ExpressionSpecification)[], unknown | ExpressionSpecification] // boolean
+    | CollatorExpressionSpecification
+    | ['format', ...(string | ['image', ExpressionSpecification] | ExpressionSpecification | {'font-scale'?: number | ExpressionSpecification, 'text-font'?: string[] | ExpressionSpecification, 'text-color': ColorSpecification | ExpressionSpecification})[]] // string
+    | ['image', unknown | ExpressionSpecification] // image
```
Comment (HarelM) [inline build/generate-style-spec.ts:153]: Why do we allow unknowns here?

File: src/style-spec/feature_filter/convert.ts [hunk 2]
```diff
@@ -61,46 +61,51 @@ export default function convertFilter(filter: FilterSpecification): unknown {
...
-        return ['!', _convertFilter(['any'].concat(filter.slice(1) as string[]) as string[], {})];
-    } else if (op === 'in') {
-        converted = convertInOp(filter[1] as string, filter.slice(2));
-    } else if (op === '!in') {
-        converted = convertInOp(filter[1] as string, filter.slice(2), true);
-    } else if (op === 'has') {
-        converted = convertHasOp(filter[1] as string);
-    } else if (op === '!has') {
-        converted = ['!', convertHasOp(filter[1] as string)];
+        converted = ['any', ...children];
```
[the remark this answers is not in the document]
Comment (cns-solutions-admin) [reply src/style-spec/feature_filter/convert.ts:91]: I didn't change the overall logic - it was and is an if-cascade.
The diff view here rather shows it as deleted all and inserted a log :-)

The changes here are only for cleaner typescript code that avoids type coercions as much as possible, e.g. for a ['any', ...LegacyFilterExpression[]] filter filter.slice(1) gives a ('any' | LegacyFilterExpression)[] (or worse), while [, conditions] = filter gives the correct LegacyFilterExpression[].

In the same way ['all', ...conditions] is a LegacyFilterExpression, while ['all'].concat(conditions) isn't.

**Comment by cns-solutions-admin** [comment]
As mentioned above, there are no real code changes, just replacing b = a.slice(1) by [, b] = a and [a].concat(b) by [a, ...b].
Additionally there were some weird type coercions "as string[]" where the elements definitely were not strings, etc.
This allowed typescript to do the type checking instead of using lots of type coercions.

The existing tests still pass.
Intersetingly I didn't find any tests testing the individual expression or filter functions!?

As to LegacyFilterExpression: that's the old (deprecated, legacy) syntax as described on https://maplibre.org/maplibre-gl-js-docs/style-spec/other/#other-filter. I decided to fully type it to make sure it was all either the old or the new syntax.

I would even think that that whole legacy filter part should be dropped from maplibre 2.

**Comment by HarelM** [comment]
2.x was already released with these legacy filters so the only proper way to discontinue them is in version 3, assuming we want to deprecate them...

**Comment by cns-solutions-admin** [comment]
They ARE deprecated since 1.0 or so (in the documentation). As to removing them, the correct version should probably be 3 (although 2.x does not feel quite stable to me yet).

**Comment by cns-solutions-admin** [comment]
Anything preventing the merge?

File: src/style-spec/feature_filter/convert.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (HarelM) [reply src/style-spec/feature_filter/convert.ts:91]: But why not change this to switch case? This particular place...

**Comment by HarelM** [comment]
I need an answer for the above code review questions :-)

File: src/style-spec/feature_filter/convert.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (cns-solutions-admin) [reply src/style-spec/feature_filter/convert.ts:91]: I don't understand - which particular place? It's a long if/else if/.../else and I didn't want to change any logic

File: build/generate-style-spec.ts (hunk 1 above)
Comment (cns-solutions-admin) [reply build/generate-style-spec.ts:153]: According to the documentation, anything is allowed here, but it is not entirely clear - obviously not arrays, as they could not be distinguished from expressions. 

So it's unknown for now, as we should not prevent anything that is allowed.

It can be improved later, but only after some expression tests have been created.

**Comment by cns-solutions-admin** [comment]
Seems, my first comment was lost or I pressed the wrong button? there were other buttons than simply "Comment"!
I didn't see your other comment to my comment. And now I don't understand it.

File: src/style-spec/feature_filter/convert.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (HarelM) [reply src/style-spec/feature_filter/convert.ts:91]: The longest else-if you can :-)
Not changing logic can be achieved by adding the relevant tests, if those don't exists let's create them (I think it also addresses the other comment you wrote) :-)

File: build/generate-style-spec.ts (hunk 1 above)
Comment (HarelM) [reply build/generate-style-spec.ts:153]: See my other comment about tests...
Since you are already deep into it I would hate to postpone it to a later time which might not come.
Let's fix it the best we can at this point in time i.e. add the relevant tests and improve the typings further. :-)

**Comment by cns-solutions-admin** [comment]
I'm doing this in the context of a project for my company, where we are evaluating, if the 3D display might be usable and useful.
So far I was not even able to compile our project with maplibre 2.x (whereas migrating from mapbox to maplibre 1.5 was fine).

I can justify to my company improving maplibre regarding bugs (e.g. wrong typings) or introducing new features we would like or need, but I do not have time for

- writing tests for those expressions which worked since mapbox 1.0 and the code of which did not change according to you
- writing tests for a conversion of deprecated filters to above expressions, which we will never use (they are deprecated!)

Sorry about that, we try to improve OSS whereever possible, but in that case I will just add some type coercions in our code and try again, or if this does not work, we will have to postpone checking out maplibre 2.x.

**Comment by HarelM** [comment]
Thanks for the honest reply. I'll take it into consideration.

File: build/generate-style-spec.ts [hunk 3]
```diff
@@ -181,7 +297,7 @@ export type CompositeFunctionSpecification<T> =
     | { type: 'interval',    stops: Array<[{zoom: number, value: number}, T]>, property: string, default?: T }
     | { type: 'categorical', stops: Array<[{zoom: number, value: string | number | boolean}, T]>, property: string, default?: T };
 
-export type ExpressionSpecificationArray = Array<unknown>;
+export type ExpressionSpecificationArray = ExpressionSpecification;
```
Comment (HarelM) [inline build/generate-style-spec.ts:300]: Why is this needed? Can we remove this export type?

File: src/style-spec/feature_filter/convert.ts [hunk 4]
```diff
@@ -9,7 +9,7 @@ type ExpectedTypes = {[_: string]: 'string' | 'number' | 'boolean'};
  * equivalent expression
  * @private
  */
-export default function convertFilter(filter: FilterSpecification): unknown {
+export default function convertFilter(filter: FilterSpecification): ExpressionFilterSpecification {
```
Comment (HarelM) [inline src/style-spec/feature_filter/convert.ts:12]: Do we really need this function? can't we simply export the inner one with a default second parameter that equals `{}`?

**Comment by HarelM** [comment]
I've merged this to a branch under the main repo.
I'll see if I can add some tests and fix the remaining comments.
Feel free to send more PRs against that branch.
